Tests/Source/ModelTester.cpp

```diff
@@ -12,95 +12,95 @@ using namespace std::chrono_literals;
 
 namespace {
   class TestAlgorithm {
-    public:
-      using Type = double;
-      using Sample = ScalarSample<double>;
-      using Arguments = typename Sample::Arguments;
+  public:
+    using Type = double;
+    using Sample = ScalarSample<double>;
+    using Arguments = typename Sample::Arguments;
 
-      template<typename ScalarView>
-      void learn(ScalarView&& reader) {
-        for(const auto& sample : reader) {
-          m_data.push_back(sample);
-        }
-        if(m_data.empty()) {
-          return;
-        }
-        compute_norm();
-        normalize_data();
+    template<typename ScalarView>
+    void learn(ScalarView&& reader) {
+      for(const auto& sample : reader) {
+        m_data.push_back(sample);
```

Comment:
If you're taking a forwarding reference to ```reader```, then you have the potential to move the ```sample``` instead of copying it. Try to take advantage of that.

Even better would be to not copy at all and operator on the ```sample``` directly.

Reply:
`ScalarView` only returns const references, so the only way to avoid copies here is by forwarding `ScalarView` itself to a data member. While possible, I am not sure whether it should be considered safe because `ScalarView` lifetime is bound to the original `Trial`, which can potentially be destroyed by the time the `Model` is trained.
The use case in this test `Model` is not really typical. It has to capture all the samples just because it computes nearest neighbors on the original sample set. Furthermore, the `ScalarSamples` are re-used on every predict, so re-computing them each time would be sub-optimal as well.
Having said that, it is more conventional to take `ScalarView` by const reference here.

Reply:
Good points, take ```ScalarView``` by ```const&```.